inmem_endpoint.go

```diff
@@ -129,3 +135,48 @@ func formatSamples(source map[string]SampledValue) []SampledValue {
 
 	return output
 }
+
+type Logger interface {
+	Warn(msg string, args ...interface{})
+}
+
+// Stream writes metrics to resp each time an interval ends. Runs until the
+// request context is cancelled.
+func (i *InmemSink) Stream(ctx context.Context, logger Logger, resp http.ResponseWriter) {
+	interval := i.getInterval()
+
+	resp.WriteHeader(http.StatusOK)
+	flusher, ok := resp.(http.Flusher)
+	if ok {
+		// call Write with 0 bytes before a flush, so that GzipResponseWriter
+		// can write response headers.
+		resp.Write([]byte(""))
+		flusher.Flush()
+	} else {
+		flusher = noopFlusher{}
```

Comment:
My understanding is that the `ResponseWriter` that come from the stdlib all support flushing, but someone could implement a wrapper that does not.

Even if it doesn't support flushing, I'm not sure that means that the metrics wont stream. I think it could mean that the writer itself will decide when to flush the bytes, instead of waiting for an explicit flush.

Reply:
Hmm I went spelunking in net/htto and it seems that if you have an HTTP 1.0 client it will use the same response object which will still be a `Flusher` but will just buffer the whole response until it's closed and then send it all.

Makes me think we probably need a time limit or something on this method? Otherwise it would be kinda easy DOS vector and a possible CVE to expose it on our clients - just issue an `GET /whatever HTTP/1.0` and don't time out and the server will sit there filling memory with metrics for as long as you keep the connection open.

Reply:
Exiting when `Flusher` is not implemented is still an option. Another option is to remove the http writer logic from this function and accept a smaller interface. The caller would then be responsible for setting up the writer/flusher.

Where do you see the unbounded buffering in memory happen?

The `InMemSink` always prunes intervals after some period, so even if we aren't consuming the stream, we will only store a fixed number of intervals.

In this function, if the write buffer fills, won't it simply block on the write? Which will stop the stream and waste a bit of memory, but won't OOM the process.

From a quick look it seemed like it used a buffer of 2048, and will call flush once that is full: https://golang.org/src/net/http/server.go#L1614 (writes to a `bufio.Writer` with size `bufferBeforeChunkingSize`).

If this were a problem, wouldn't it be true for everything that uses the `net/http` server? Shouldn't the docs call this out very loudly that you must explicitly flush large responses? I don't see any such warnings, and the docs say:

>  Once the headers have been flushed (due to either an explicit Flusher.Flush call or writing enough data to trigger a flush).

Which again implies to me that flush will happen anyway, even when not explicitly called.

Reply:
I'm not sure honestly. 

I think the issue I see is that I can't find any documentation about how `Flusher` works when the underlying request is HTTP 1.0 which doesn't support chunked encoding. 

What I _did_ find was net/http code that will not try to send chunked content in that case (because it can't) but will just wait until the end of the request and send the whole thing at once since it needs a content length to put in the header before it can begin writing the body to the wire.

So I'm not sure how that would interact with this handler - this handler will loop indefinitely as long as the client doesn't close their connection and will continue trying to write more data all that time, but the server will be waiting for it to return before it sends any of it.

You're probably right that it will fill a buffer and then block trying to write more which perhaps limits the impact, but it does mean that request will "leak" that blocked goroutine indefinitely until the client goes away maybe 🤷 .

I think this is different from most "normal" web serving because you typically don't have an _unbounded_ length. Even transferring a huge file from disk to an HTTP 1.0 client is fine since you know the content length ahead of time so can send the header and start streaming the file in chunks.

That said you're right that HTTP 1 client could cause issues with large writes made using net/http without explicitly setting the content length and it's surprising to me that the behaviour in that case is not better documented if it's actually bad, but it's also not clear it's not bad!

We could test it out! Or we could take the easier path of just refusing to server HTTP 1.0 requests because streaming doesn't actually work for them anyway (and no reasonable client we care about would be stuck on 1.0) and then we don't have to worry about the exact behaviour at all. What do you think?

Reply:
This is the code I was referring to btw. There is no separate 1.0 implementation, just conditionals in the main server implementation: https://golang.org/src/net/http/server.go#L1436

I didn't trace it all through fully though so maybe it's all magically fine somehow, I just don't see how based on how I think HTTP 1.0 works (requires a Content length header up front).

Reply:
Oh and to be clear, the `Flusher` not being implemented part doesn't even help here since even for 1.0 requests the same response type is used that _does_ implement flusher so that's not significant either way.

Reply:
Ok, I tried it out, using this test binary: https://github.com/dnephin/go-metrics/commit/3ee71ae6513150d0bcc1a6f574187a22a7945da5

It runs a few goroutines to produce metrics, and runs an `http.Server` with this single endpoint.

I performed an HTTP request using:  
```
curl -v --http1.0 http://localhost:8080/
```

The result was exactly what I would expect. I see the metrics being logged to stdout of curl every `interval` seconds (10s).

Reply:
Yeah, I think that is sufficient.

I just ran a simple test myself to confirm exactly what was going on.

I misremembered the HTTP 1 spec. I thought that since chunked encoding wasn't supported, the only option was to declare the response body length in headers up front.

It turns out that although it's not clear from the RFC section on Content-Length (which I read earlier too), it's valid to not return a content length header and allow the client to assume the body continues until EOF. https://stackoverflow.com/questions/15991173/is-the-content-length-header-required-for-a-http-1-0-response/15995101#:~:text=59-,Is%20the%20Content%2DLength%20header,for%20a%20HTTP%2F1.0%20response%3F&text=A%20valid%20Content%2DLength%20field,messages%20containing%20an%20entity%20body.

So yeah it will always write it immediately 😌 . Sorry for the misdirection!